### PR TITLE
Feat/oracle router

### PR DESCRIPTION
**Pull request by jakeal-papaca**
## Description
 - OracleRouter Contract

## Why
Because we are integrating Chainlink.

### CODE REVIEW

File: contracts/6/protocol/OracleRouter.sol [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    address[] calldata token0s,
+    address[] calldata token1s,
+    uint256[] calldata maxPriceDeviationList,
+    PriceOracle[][] calldata allSources
+  ) external onlyOwner {
+    require(
+      token0s.length == token1s.length &&
+        token0s.length == allSources.length &&
+        token0s.length == maxPriceDeviationList.length,
+      "inconsistent length"
```
Comment (spicysquid168) [inline contracts/6/protocol/OracleRouter.sol:73]: Use `contract::function:: revert_msg" format for revert message

File: contracts/6/protocol/OracleRouter.sol [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  /// @param sources Oracle sources for the token pair
+  function _setPrimarySources(
+    address token0,
+    address token1,
+    uint256 maxPriceDeviation,
+    PriceOracle[] memory sources
+  ) internal {
+    require(
+      maxPriceDeviation >= MIN_PRICE_DEVIATION && maxPriceDeviation <= MAX_PRICE_DEVIATION,
+      "bad max deviation value"
```
Comment (spicysquid168) [inline contracts/6/protocol/OracleRouter.sol:93]: Use `contract::function:: revert_msg" format for revert message

File: contracts/6/protocol/OracleRouter.sol [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    emit SetPrimarySources(token0, token1, maxPriceDeviation, sources);
+  }
+
+  /// @dev Return token0/token1 price
+  /// @param token0 Token0 to get price of
+  /// @param token1 Token1 to get price of
+  /// NOTE: Support at most 3 oracle sources per token
+  function _getPrice(address token0, address token1) public view returns (uint256) {
+    uint256 candidateSourceCount = primarySourceCount[token0][token1];
+    require(candidateSourceCount > 0, "no primary source");
```
Comment (spicysquid168) [inline contracts/6/protocol/OracleRouter.sol:113]: Use `contract::function:: revert_msg" format for revert message

File: contracts/6/protocol/OracleRouter.sol [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    //     --> else revert
+    // - 3 valid sources --> check deviation threshold of each pair
+    //     --> if all within threshold, return median
+    //     --> if one pair within threshold, return average of the pair
+    //     --> if none, revert
+    // - revert otherwise
+    if (validSourceCount == 1) {
+      return prices[0]; // if 1 valid source, return
+    } else if (validSourceCount == 2) {
+      require(prices[1].mul(1e18) / prices[0] <= maxPriceDeviation, "too much deviation (2 valid sources)");
```
Comment (spicysquid168) [inline contracts/6/protocol/OracleRouter.sol:144]: Use `contract::function:: revert_msg" format for revert message

File: contracts/6/protocol/OracleRouter.sol [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+      bool midMinOk = prices[1].mul(1e18) / prices[0] <= maxPriceDeviation;
+      bool maxMidOk = prices[2].mul(1e18) / prices[1] <= maxPriceDeviation;
+      if (midMinOk && maxMidOk) {
+        return prices[1]; // if 3 valid sources, and each pair is within thresh, return median
+      } else if (midMinOk) {
+        return prices[0].add(prices[1]) / 2; // return average of pair within thresh
+      } else if (maxMidOk) {
+        return prices[1].add(prices[2]) / 2; // return average of pair within thresh
+      } else {
+        revert("too much deviation (3 valid sources)");
```
Comment (spicysquid168) [inline contracts/6/protocol/OracleRouter.sol:156]: Use `contract::function:: revert_msg" format for revert message

File: contracts/6/protocol/OracleRouter.sol [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+      }
+    } else {
+      revert("more than 3 valid sources not supported");
```
Comment (spicysquid168) [inline contracts/6/protocol/OracleRouter.sol:159]: Use `contract::function:: revert_msg" format for revert message

File: contracts/6/protocol/OracleRouter.sol [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    require(candidateSourceCount > 0, "no primary source");
+    uint256[] memory prices = new uint256[](candidateSourceCount);
+    // Get valid oracle sources
+    uint256 validSourceCount = 0;
+    for (uint256 idx = 0; idx < candidateSourceCount; idx++) {
+      try primarySources[token0][token1][idx].getPrice(token0, token1) returns (uint256 price, uint256 lastUpdate) {
+        prices[validSourceCount++] = price;
+      } catch {}
+    }
+    require(validSourceCount > 0, "no valid source");
```
Comment (spicysquid168) [inline contracts/6/protocol/OracleRouter.sol:122]: Use `contract::function:: revert_msg" format for revert message

File: contracts/6/protocol/OracleRouter.sol [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    for (uint256 i = 0; i < validSourceCount - 1; i++) {
```
Comment (spicysquid168) [inline contracts/6/protocol/OracleRouter.sol:123]: are you sorting the prices here? If so please up the comment here too.

File: test/OracleRouter.test.ts [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+            [BigNumber.from('900000000000000000'), BigNumber.from('1000000000000000000').div(9)],
+          )
+
+          await oracleRouterAsDeployer.setPrimarySources(
+            token0.address,
+            token1.address,
+            BigNumber.from('1100000000000000000'),
+            [simplePriceOracle.address, bobPriceOracle.address],
+          )
+          await expect(oracleRouterAsAlice.getPrice(token0.address, token1.address)).to.reverted
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:269]: assert revert message too

File: test/OracleRouter.test.ts [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+            [BigNumber.from('800000000000000000'), BigNumber.from('1000000000000000000').div(8)],
+          )
+
+          await oracleRouterAsDeployer.setPrimarySources(
+            token0.address,
+            token1.address,
+            BigNumber.from('1100000000000000000'),
+            [simplePriceOracle.address, bobPriceOracle.address, evePriceOracleAsFeeder.address],
+          )
+          await expect(oracleRouterAsAlice.getPrice(token0.address, token1.address)).to.reverted
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:321]: assert revert message

File: test/OracleRouter.test.ts [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+        })
+      })
+      context('when source1 and source2 has too much deviation', async () => {
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:324]: I think this case is "when price0 and price1 are within max deviation, but price 2 doesn't"

File: test/OracleRouter.test.ts [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+          )
+
+          await oracleRouterAsDeployer.setPrimarySources(
+            token0.address,
+            token1.address,
+            BigNumber.from('1200000000000000000'),
+            [simplePriceOracle.address, bobPriceOracle.address, evePriceOracleAsFeeder.address],
+          )
+          const [price, lastTime] = await oracleRouterAsAlice.getPrice(token0.address, token1.address)
+          expect(price).to.eq(BigNumber.from('850000000000000000'))
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:349]: Please add comment here where does `850000000000000000` come from

File: test/OracleRouter.test.ts [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+        })
+      })
+      context('when source2 and source3 has too much deviation', async () => {
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:352]: this one should be price1 and price2 are within max deviation, but price0 doesn't?

File: test/OracleRouter.test.ts [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+          )
+
+          await oracleRouterAsDeployer.setPrimarySources(
+            token0.address,
+            token1.address,
+            BigNumber.from('1200000000000000000'),
+            [simplePriceOracle.address, bobPriceOracle.address, evePriceOracleAsFeeder.address],
+          )
+          const [price, lastTime] = await oracleRouterAsAlice.getPrice(token0.address, token1.address)
+          expect(price).to.eq(BigNumber.from('950000000000000000'))
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:377]: same as above, please show your logic where does you find `950000000000000000`

File: test/OracleRouter.test.ts [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+        })
+      })
+      context('when source1, source2 and source3 is ok', async () => {
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:380]: are ok

Comment (spicysquid168) [reply test/OracleRouter.test.ts:380]: I think price is better word that source here as we compare prices not sources

File: test/OracleRouter.test.ts [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+          )
+
+          await oracleRouterAsDeployer.setPrimarySources(
+            token0.address,
+            token1.address,
+            BigNumber.from('1200000000000000000'),
+            [simplePriceOracle.address, bobPriceOracle.address, evePriceOracleAsFeeder.address],
+          )
+          const [price, lastTime] = await oracleRouterAsAlice.getPrice(token0.address, token1.address)
+          expect(price).to.eq(BigNumber.from('900000000000000000'))
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:405]: Add comment to explain where does this '900000000000000000' come from?

File: test/OracleRouter.test.ts [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+                simplePriceOracle.address,
+                simplePriceOracle.address,
+              ],
+            ),
+          ).to.revertedWith('sources length exceed 3')
+        })
+      })
+      context('when successfully', async () => {
+        it('should successfully', async () => {
+          await oracleRouterAsDeployer.setPrimarySources(
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:137]: assert event emit too

File: test/OracleRouter.test.ts [hunk 18]
```diff
@@ -0,0 +1,410 @@
...
+            [token1.address, token3.address],
+            [BigNumber.from('1000000000000000000'), BigNumber.from('900000000000000000')],
+            [[simplePriceOracle.address]],
+          ),
+        ).to.revertedWith('inconsistent length')
+      })
+    })
+    context('when successfully', async () => {
+      it('should successfully', async () => {
+        await oracleRouterAsDeployer.setMultiPrimarySources(
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:189]: assert event emit too

File: test/OracleRouter.test.ts [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+          [token0.address, token2.address],
+          [token1.address, token3.address],
+          [BigNumber.from('1000000000000000000'), BigNumber.from('1100000000000000000')],
+          [[simplePriceOracle.address], [simplePriceOracle.address, bobPriceOracle.address]],
+        )
+        const sourceT0T1 = await oracleRouterAsDeployer.primarySources(token0.address, token1.address, 0)
+        const sourceCountT0T1 = await oracleRouterAsDeployer.primarySourceCount(token0.address, token1.address)
+        const maxPriceDeviationT0T1 = await oracleRouterAsDeployer.maxPriceDeviations(token0.address, token1.address)
+        const source0T2T3 = await oracleRouterAsDeployer.primarySources(token2.address, token3.address, 0)
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:198]: source0T2T3 -> sourceT2T3

File: test/OracleRouter.test.ts [hunk 20]
```diff
@@ -0,0 +1,410 @@
...
+    })
+    context('when successfully', async () => {
+      it('should successfully', async () => {
+        await oracleRouterAsDeployer.setMultiPrimarySources(
+          [token0.address, token2.address],
+          [token1.address, token3.address],
+          [BigNumber.from('1000000000000000000'), BigNumber.from('1100000000000000000')],
+          [[simplePriceOracle.address], [simplePriceOracle.address, bobPriceOracle.address]],
+        )
+        const sourceT0T1 = await oracleRouterAsDeployer.primarySources(token0.address, token1.address, 0)
```
Comment (spicysquid168) [inline test/OracleRouter.test.ts:195]: should assert invest pair too i.e. sourceT1T0 and etc.

Comment (spicysquid168) [reply test/OracleRouter.test.ts:195]: please leave space or comment between each set of pair, so more readable

**Review by spicysquid168 — DISMISSED** [review]

File: contracts/6/protocol/OracleRouter.sol [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+        return prices[1]; // if 3 valid sources, and each pair is within thresh, return median
+      } else if (midMinOk) {
+        return prices[0].add(prices[1]) / 2; // return average of pair within thresh
+      } else if (maxMidOk) {
+        return prices[1].add(prices[2]) / 2; // return average of pair within thresh
+      } else {
+        revert("OracleRouter::getPrice:: too much deviation 3 valid sources");
+      }
+    }
+  }
```
Comment (abraham-alpaca) [inline contracts/6/protocol/OracleRouter.sol:153]: can be sth like
```
if (validSourceCount == 1) return prices[0];
if (validSourceCount == 2) {
     require(prices[1].mul(1e18) / prices[0] &lt;= maxPriceDeviation,       
     "OracleRouter::getPrice:: too much deviation 2 valid sources"      );     
      return prices[0].add(prices[1]) / 2; // if 2 valid sources, return average
}
 bool midP0P1Ok = prices[1].mul(1e18) / prices[0] &lt;= maxPriceDeviation;     
 bool midP1P2Ok = prices[2].mul(1e18) / prices[1] &lt;= maxPriceDeviation;
 if (midP0P1Ok && midP1P2Ok) return prices[1]; // if 3 valid sources, and each pair is within thresh, return median
 if (midP0P1Ok)  return prices[0].add(prices[1]) / 2; // return average of pair within thresh
 if (midP1P2Ok) return prices[1].add(prices[2]) / 2; // return average of pair within thresh
 revert("OracleRouter::getPrice:: too much deviation 3 valid sources");

```

**Review by abraham-alpaca — APPROVED** [review]

**Review by spicysquid168 — APPROVED** [review]

**Pull request merged**